### DISCUSSION TITLE
Data fetcher interface generation

### DISCUSSION THREAD

**Discussion by zaenk** [Netflix/dgs-framework#234] [Ideas]
There is one feature I'm missing from the current code generation, which is a similar option to [OpenAPI Spring generator's](https://openapi-generator.tech/docs/generators/spring/) `interfaceOnly` flag. This would generate the data fetchers as interfaces with the required `@DgsData`, `@InputArgument` etc annotations.

This would enable generating artifacts with only DTO implementations and API interfaces, decoupling the API specification and code generation part from the data fetcher implementation.

I've already started poking around the projects to see what's possible right now:
- in the framework there is one roadblock, that is `@Dsg*` annotations are not picked up from implemented interfaces. You can see this [demonstrated here](https://github.com/zaenk/dgs-api-first-datafetcher/pull/2/files), and a possible [~fix here~](https://github.com/zaenk/dgs-framework/commit/fe5d1ccdd3382fa1cddf2010f81c1c18f37ac202) https://github.com/Netflix/dgs-framework/pull/320
- option for the generator to generate interfaces rather the stubs as examples. It could be a new set of stuff generated or users could just set the examples target path the the generated sourcea path. I already went ahead to see if I could make it work [~here~](https://github.com/zaenk/dgs-codegen/commit/62d12d6b6f3e7f932d4e539806be72858a317332) https://github.com/Netflix/dgs-codegen/pull/126

It seems to me that these could be turned into a workable solution in a short time, but I would want to get some input first on the matter.

There's a few things to consider before I'd go forward and finish up the previous two branches for a PR.

One is the list of arguments, whether is it enough to only generate the arguments defined in the schema or some other, or should other stuff, like `DataFetchingEnvironment` and other should always be added.

The other is the programming model. At the current state, DGS supports Spring Web MVC and plans to also support Flux. Just generating the interface and using it in a similar way as the OpenAPI generated interfaces could work in the short term, but might not be optimal later with flux.

Looking at Expedia's library method/function arguments are the preferred way to pass dependencies to data fetchers. While this doesn't seem to be a requirement by Spring Flux, this makes me wonder, whether generating interface methods/functions with only the input arguments from the schema might not be enough.

I don't have much experience with GraphQL nor with Flux, and I probably would use it with Web MVC anyway, but I'm open to suggestions, or whether is this something that others would find useful.

edit: linked related PRs

**Comment by paulbakker** [comment]
So what you would want to achieve is manually implement those datafetcher interfaces, but without using any of its annotations? Why is that better? It _looks_ like you're then decoupled from the framework with the code you write, but in reality that's not the case. You would still follow a code structure that's quite specific to GraphQL, and things like data loaders.

Can you share more about the benefits you see when we would have this? Maybe illustrate it with some use cases?

**Reply by zaenk** [reply]
Yes, decoupling from the framework is not my intent here.

What I have in mind is that I'd like to keep the schema definitions in a separate repository and also add some tooling in there for code generation. This tooling would publish the generated code as api libraries, e.g. it would generate and publish a `foo-bar-android-client.jar`, `foo-bar-angular-client` npm package and `foo-bar-kotlin-server.jar`.

In this setup, I'd expect that the generated server jar would contain the data classes (these are generated right now) and some interfaces describing the queries and mutations that were defined in the schema. Things like these are not generated right now for kotlin, and for java only really basic examples are generated.

But let's forget the different repository part for now and focus on code generation.

I'd like to achieve something similar to this: https://github.com/zaenk/specification-first-with-openapi-example . Here the generated sources take care of adding the right annotations (mapping http methods to handlers), constants (mapping url paths to handlers) and input-output types.

For DGS I'd imagine the generated interface would look like this:

```kotlin
interface ShowDataFetcher {
  @DgsData(parentType = DgsConstants.QUERY_TYPE, field = DgsConstants.QUERY.Shows) // this annotation is not detected by the framework right now
  fun shows(@InputArgument("titleFilter") titleFilter: String?): List<Show>
}
```

And the implementation:

```kotlin
@DgsComponent // need this here for component scanning
class ShowDataFetcherImpl(
  private val showService: ShowService
) : ShowDataFetcher {
  override fun shows(titleFilter: String?): List<Show> =
    titleFilter?.let { title -> showService.shows().filter { it.title.contains(title) } } ?: showService.shows()
}
```

*edit: adds missing implements clause to example*
